src/Microsoft.Restier.Publishers.OData/Model/RestierModelBuilder.cs

```diff
@@ -49,7 +55,19 @@ class RestierModelBuilder : IModelBuilder
 
             context.EntitySetTypeMapCollection.Clear();
 
+            ExtendModel(builder);
+
             return builder.GetEdmModel();
         }
+
+        /// <summary>
+        /// This method allow user to extend the model with more entity type and entity set before operation model builder starts
+        /// A sub class need to be implemented and register as DI service like "AddTransient&lt;RestierModelBuilder, CustomizedRestierModelBuilder&gt;"
+        /// </summary>
+        /// <param name="builder"></param>
+        public virtual void ExtendModel(ODataConventionModelBuilder builder)
```

Comment:
So, I think you guys are making this *entirely* too complicated, partially because you are implementing DI incorrectly. The point is to allow the end-user to manipulate the model that gets used by RESTier. Nearly 100% of users will be used to using the `ODataConventionModelBuilder`, because that is what is used by WebAPI. I think you would be hard-pressed to find anyone who **wants** to manipulate the result of the `builder.GetEdmModel()` call. If that is the case, than this function is returning entirely the wrong object type.

Further, the point of Dependency Injection is to create base classes that can be inherited from and used, not to create chains that have to be implemented in specific orders... that can be done without DI. All of the "chain of service" stuff is overcomplicating everything about how the end-developer interacts with the service. It's totally counter-intuitive. I know this because I spent 12 hours over the weekend trying to implement some of this stuff in an API, and it is *very* confusing.

I personally think the design should look like this:

```cs
///<summary>
/// Handles the automagic process of constructing a ModelBuilder instance from a given DbContext.
///</summary>
///<remarks>
/// Inherit from this class, override BuildModelAsync, and call base.BuildModelAsync() to customize your model.
///</remarks>
public class RestierModelBuilder : IModelBuilder 
{
    ///<summary>
    /// The ODataConventionModelBuilder instance that will contain the generated model.
    ///</summary>
    public ODataConventionModelBuilder ModelBuilder { get; set; }

    public RestierModelBuilder(string namespace)
    {
        ModelBuilder = new ODataConventionModelBuilder();
        ModelBuilder.Namespace = namespace;
    }

    ///<summary>Please start documenting these methods so that Intellisense works.</summary>
    public async Task BuildModelAsync(ModelContext context, CancellationToken cancellationToken)
    {
        ModelBuilder.YaddaYaddaYadda....
    }

    public IEdmModel GetEdmModel() => return ModelBuilder.GetEdmModel(); 

}
```
Then you don't need virtual methods, you don't need chains, you don't need anything unusual... just register your custom inherited ModelBuilder with the DI container before you call `MapRestierRoute()` in `WebApiConfig.cs`.

IMO, **any** place where you are having to call InnerXXX is a candidate for this type of cleaner, simplified architecture. And with this structure, 100% of the code used to build existing WebAPI OData models will be compatible with RESTier.

HTH!